### PR TITLE
Nginx bug fixes and integration tests

### PR DESCRIPTION
**Pull request by kuba**
More info about nature of changes in the commit messages. In short: integration testing of nginx plugin in Travis!

@diracdeltas, could you please review? I hope it doesn't interfere too much with your work... :)

### CODE REVIEW

File: letsencrypt_nginx/configurator.py [hunk 1]
```diff
@@ -263,6 +268,18 @@ def get_all_names(self):
...
+        cert_pem = crypto_util.make_ss_cert(
+            key.pem, domains=[socket.gethostname()])
+        cert = os.path.join(tmp_dir, "cert.pem")
+        with open(cert, 'w') as cert_file:
+            cert_file.write(cert_pem)
+        return cert, key.file
+
     def _make_server_ssl(self, vhost):
         """Makes a server SSL based on server_name and filename by adding
         a 'listen 443 ssl' directive to the server block.
```
Comment (diracdeltas) [inline letsencrypt_nginx/configurator.py]: Docstring should reflect the fact that the port is configurable.

File: letsencrypt_nginx/configurator.py [hunk 2]
```diff
@@ -274,17 +291,22 @@ def _make_server_ssl(self, vhost):
         :type vhost: :class:`~letsencrypt_nginx.obj.VirtualHost`
 
         """
-        ssl_block = [['listen', '443 ssl'],
-                     ['ssl_certificate',
-                      '/etc/ssl/certs/ssl-cert-snakeoil.pem'],
-                     ['ssl_certificate_key',
-                      '/etc/ssl/private/ssl-cert-snakeoil.key'],
+        snakeoil_cert, snakeoil_key = self._get_snakeoil_paths()
+        ssl_block = [['listen', '{0} ssl'.format(self.config.dvsni_port)],
```
Comment (diracdeltas) [inline letsencrypt_nginx/configurator.py]: Should `config.dvsni_port` be changed to 443 in `deploy_cert`? aka, do we expect that the user wants to enable SSL on the same port that is used for DVSNI challenges?

Comment (kuba) [inline letsencrypt_nginx/configurator.py]: That's a good point. Using `config.dvsni_port` everywhere was the fastest way to get integration testing working. I think we should also have `--nginx-ssl-vhost-port` (default `443`), but hardcoding must be avoided. Follow-up PR?

File: letsencrypt_nginx/configurator.py (hunk 1 above)
Comment (kuba) [inline letsencrypt_nginx/configurator.py]: ups, forgot about this one, thanks!

**Comment by diracdeltas** [comment]
Haven't tested this, but the code looks fine. I'm not working on anything with Nginx currently (am busy with other job) so this doesn't conflict. Thanks!

**Comment by kuba** [comment]
Thanks for the review! :) Fixed the docsting, left port manipulation for a follow-up PR.

Tested locally, and against demo server (Debian jessie, fresh nginx config). Unfortunately, this has been hit again by https://github.com/letsencrypt/boulder/issues/420 (https://travis-ci.org/letsencrypt/letsencrypt/jobs/68617740#L1258)

**Comment by jdkasten** [comment]
Thanks!

File: letsencrypt_nginx/configurator.py (hunk 2 above)
Comment (jdkasten) [inline letsencrypt_nginx/configurator.py]: I think this goes into the general category of finding the correct virtualhost.  I recently added a Vhost selection screen for Apache. https://github.com/letsencrypt/letsencrypt/blob/master/letsencrypt_apache/display_ops.py that allows the user to choose a vhost when there isn't a clear selection.  If it is of any use, I will move it into a common file within the core.

Note: that the vhost selection does not solve the problem of when you have to upgrade a preexisting vhost.  That would still require the ```--ssl-vhost-port``` option

**Pull request merged**